### ISSUE TITLE
[Breaking Change Request] [io/http] Drop Content-Length:0 header from CONNECT,DELETE,GET,HEAD requests

### ISSUE THREAD

**Issue by aam** [dart-lang/sdk#45871]
Summary
---
We propose to not to include `Content-Length` header in CONNECT,DELETE,GET,HEAD http requests when its value is `0`(when there is no body as part of http request).

Rationale
---
Presence of `Content-Length:0` header runs against guidelines provided in RFC-7320 https://tools.ietf.org/html/rfc7230#section-3.3.2, which states that 
```
A user agent SHOULD NOT send a
   Content-Length header field when the request message does not contain
   a payload body and the method semantics do not anticipate such a body.
```

The fact that dart:io `HttpClientRequest` code includes `Content-Length:0` header in `GET` requests was recently found to be a culprit of inability for Dart http client code to talk to Google Cloud Run websocket services (see https://github.com/dart-lang/sdk/issues/45139). So this change will fix that problem.

See https://dart-review.googlesource.com/c/sdk/+/194881 for the actual change.

**Comment by aam** [comment]
@franklinyow, just wanted to let you know that announcement email has been sent out (https://groups.google.com/a/dartlang.org/g/announce/c/b0cncgjcRf4)

**Comment by rspilker** [comment]
What about other methods that also don't have defined semantics for a payload, like [`DELETE`](https://tools.ietf.org/html/rfc7231#section-4.3.5)?

**Comment by aam** [comment]
> @rspilker wrote
>
> What about other methods that also don't have defined semantics for a payload, like [`DELETE`](https://tools.ietf.org/html/rfc7231#section-4.3.5)?

Good question, thanks for pointing that out.
Per https://tools.ietf.org/html/rfc7231 It looks like DELETE and CONNECT in addition to GET and HEAD have similar language that "A payload within a ... request message has no defined semantics;
   sending a payload body on a HEAD request might cause some existing
   implementations to reject the request."
I believe we should extend proposed change to all four.

**Comment by aam** [comment]
any updates on this @franklinyow ?

**Comment by franklinyow** [comment]
Ping @vsmenon @matanlurey @Hixie for approval

**Comment by Hixie** [comment]
Following the standards SGTM.

**Comment by vsmenon** [comment]
lgtm

**Comment by franklinyow** [comment]
Approved

**Comment by daniel-mf** [comment]
Any updates?
Really would like to have this fixed so I could remove the massive workaround in my project.
Thanks!

**Comment by thomas-ferchau** [comment]
If I read the information in the [commit](https://github.com/dart-lang/sdk/commit/6c254fab89adbeebfb616cdec497260db4cb7cf7) correctly, the change is included in the dev releases since 2.14.0-181.0.dev

**Comment by aam** [comment]
Fixed in https://github.com/dart-lang/sdk/commit/6c254fab89adbeebfb616cdec497260db4cb7cf7
